### DISCUSSION TITLE
this.changes.length is empty when using tfs git, but it's there when using VSTFS. On Premise Azure Devops 2019

### DISCUSSION THREAD

**Discussion by mlop3s** [rfennell/AzurePipelines#961] [Q&A] [answered]
Hi,

i'm generating the release notes using the handlebars version. The extension version is 3.44.4.
On the project using VSTFS I get the this.changes.length > 0 and I can itereate through the files.

On the Git version with Pull Request based commits, the this.changes.length is always 0.

Perhaps the extension is not getting the commits? (GET ... /git/repositories/{repositoryId}/commits )
But using the (GET ... /_apis/tfvc/changesets)

Or the repositoryId is somehow missing?

Both projects is using AzureDevops 2019 On Premise.

This is the template:
```
{{#forEach commits}}
{{#if isFirst}}### Associated commits  (only shown if CS) {{/if}}
* **ID {{this.id}}** 
  -  **Message:** {{get_only_message_firstline this.message}}
  -  **Commited by:** {{this.author.displayName}} 
   -  **FileCount:** {{this.changes.length}} 
{{/forEach}}
```

Any help appreciated.

Cheers,
Marco

**Comment by rfennell** [comment, marked as answer]
Most users of this task are using Git, so I don't think there is a fundamental issue.

Can you provide a detailed log, rerun the pipeline with the variable `system.debug=true` or tick the `run with diagnostic' on the start run dialog. This will show all the logic used get the details.

**Reply by mlop3s** [reply]
Hi,

sorry it took so long.
I guess this is the problem:

2021-03-06T10:08:46.8554348Z Enriched change bac5ca5e8a366c9e741deaaf225791642df51a72 of type TfsGit
2021-03-06T10:08:46.8557626Z Cannot enriched as location URL not in dev.azure.com or xxx.visualstudio.com format

Not sure why it's expecting this url format.
On Premise has no restrictions on the url name.

On the payload i can see the changes:
``` json
"changes":{"href":"http://my-own-url:8080/Collection/cb4c301f-ea68-4751-8baa-8f9346f1a21d/_apis/git/repositories/4dd8dca6-859f-48bc-bf10-52c95b53fc22/commits/1412b3399667080a1d36f0a044beb10e15c25bb2/changes"}},"push":{"pushedBy":{"displayName":"OneUSer... 
```

This is how the changes look like when navigating to the url:

```
{
	"changeCounts": {
		"Edit": 8
	},
	"changes": [
		{
			"item": {
				"objectId": "12f61400947a79905cfd03de121bdaa4bcbda603",
				"originalObjectId": "e94396a4d2ad00ec1cd0712c142470c19316a410",
				"gitObjectType": "tree",
				"commitId": "1412b3399667080a1d36f0a044beb10e15c25bb2",
				"path": "/MyProduct",
				"isFolder": true,
				"url": "http://my-own-url:8080/Collection/cb4c301f-ea68-4751-8baa-8f9346f1a21d/_apis/git/repositories/4dd8dca6-859f-48bc-bf10-52c95b53fc22/items/MyProduct?versionType=Commit&version=1412b3399667080a1d36f0a044beb10e15c25bb2"
			},
			"changeType": "edit"
		},
		{
			"item": {
				"objectId": "af1d45c2dead913280f76240a2d4c0530e90e7ab",
				"originalObjectId": "297caefbbfa06c7cdb8d1829c36927d89109cadd",
				"gitObjectType": "tree",
				"commitId": "1412b3399667080a1d36f0a044beb10e15c25bb2",
				"path": "/MyProduct/Win32",
				"isFolder": true,
				"url": "http://my-own-url:8080/Collection/cb4c301f-ea68-4751-8baa-8f9346f1a21d/_apis/git/repositories/4dd8dca6-859f-48bc-bf10-52c95b53fc22/items/MyProduct%2FWin32?versionType=Commit&version=1412b3399667080a1d36f0a044beb10e15c25bb2"
			},
			"changeType": "edit"
		},
		{
			"item": {
				"objectId": "362ae496a428619b0f162025991abe3177a748a1",
				"originalObjectId": "9fca542aed6443fdf0e39bcd3572c80806ba39c1",
				"gitObjectType": "blob",
				"commitId": "1412b3399667080a1d36f0a044beb10e15c25bb2",
				"path": "/MyProduct/Win32/ViewModel.cs",
				"url": "http://my-own-url:8080/Collection/cb4c301f-ea68-4751-8baa-8f9346f1a21d/_apis/git/repositories/4dd8dca6-859f-48bc-bf10-52c95b53fc22/items/MyProduct/ViewModel.cs?versionType=Commit&version=1412b3399667080a1d36f0a044beb10e15c25bb2"
			},
			"changeType": "edit"
		},
```

It looks like it's filtered out. Perhaps the change format above is not as expected?

Let me know if you still need the whole log.
Thanks again for the support.

Cheers,
Marco

**Reply by rfennell** [reply]
You are suffering from an issue that my extensions are not tested that extensively with on-premise, they are very much cloud-first these days as the data I get show that is where the vast major of the consume them.

I will need to make a code change to allow on-premise URLs too.

I have created a new issue to track this #965

**Reply by rfennell** [reply]
I have made a start on a fix to remove this unnecessary domain check. It has passed my regression tests, but these are cloud focused. I need to provision an on-prem server for further testing.

However, @mlop3s  would you be able to pull down the new version before it is published to the marketplace? You can find it in my public [Azure DevOps project I used for CI/CD. t the build 3.47.1](https://dev.azure.com/richardfennell/GitHub/_build/results?buildId=8941&view=results). The pipeline is paused at the manual step while I do this extra testing

**Reply by rfennell** [reply]
Turns out I had a stored Azure DevOps Server 2019 test VM. So I was able to test the change and now am able to see the enrichment of commits

```
Getting items associated with only the current build
Get the file details associated with the commits
Enriched change f5b39f1a11ec2e703cbeb86b8e77be90956de11d of type TfsGit
Enriched with details of 1 files
Get any test details associated with the build
```

I will get this update pushed to the marketplace. The release version, after the PR to main will be 3.47.3

**Reply by mlop3s** [reply]
Great. Thanks. I will test it asap.
Let me know if i can help any further with the project js, c#, c++ or testing.

**Reply by mlop3s** [reply]
Working now:

ID02f790a7c437fa6b2360cd87bac2aa0afd01fde7

Message: Merged PR 263: preparing ...
Commited by: mlop3s
**FileCount: 9**

You may :shipit:  👍 

Thanks.
